roboteam_ai/src/coach/midField/HarassRobotCoach.cpp

```diff
@@ -0,0 +1,270 @@
+            return harassTarget;
+            //position
+        } else {
+            return findRobotToHarass(thisRobot, myIndex, false);
+        }
+    }
+}
+
+int HarassRobotCoach::getRobotIndexCloseToEnemyRobot(const world::World::RobotPtr &enemyRobot) const {
+    double closestDistanceSquared = 9e9;
+    int bestIndex = - 1;
+    for (unsigned int i = 0; i < currentRobotPositions.size(); i ++) {
+        auto &robotPos = currentRobotPositions[i];
+        double lengthSquared = (robotPos - enemyRobot->pos).length2();
+        if (lengthSquared < closestDistanceSquared) {
+            closestDistanceSquared = lengthSquared;
+            bestIndex = i;
+        }
+    }
+    return bestIndex;
+}
+
+Vector2 HarassRobotCoach::harassRobot(int myIndex, int id) {
+    Vector2 target;
+    RobotPtr robotToHarass = world::world->getRobotForId(id, false);
+
+    if (world::world->theirRobotHasBall(id)) {
+        target = ball->pos;
+    } else {
+
+        if (robotToHarass->vel.length() > MINIMUM_HARASS_VELOCITY) {
+            Vector2 harassingDistance = {robotToHarass->vel.length()*HARASSER_SECONDS_AHEAD, 0};
+            target = robotToHarass->pos + harassingDistance.rotate(robotToHarass->vel.toAngle());
+        }
+        else {
+            target = robotToHarass->pos;
+            target.x -= DEFAULT_HARASSING_DISTANCE;
+        }
+
+        Vector2 projection = currentRobotPositions[myIndex].project(robotToHarass->pos, target);
+        if ((projection - currentRobotPositions[myIndex]).length() < 0.05) {
+            target = robotToHarass->pos;
+        }
+    }
+
+    return target;
+}
+
+HarassRobotCoach::HarassTarget HarassRobotCoach::initialize(const Vector2 &currentLocation, int &myIndex) {
+    Vector2 target = {bestXPos, currentLocation.y};
+    currentRobotPositions.push_back(currentLocation);
+    targetRobotPositions.push_back(target);
+    targetRobotsToHarass.push_back(RobotPtr(nullptr));
+    myIndex = currentRobotPositions.size() - 1; // set my index (side-effect, but works well tm)
+
+    HarassTarget harassTarget;
+    harassTarget.harassPosition = target;
+    harassTarget.harassRobot = -1;
+
+    return harassTarget;
+}
+
+Angle HarassRobotCoach::getHarassAngle(const HarassRobotCoach::RobotPtr &thisRobot, int &myIndex) {
+    ball = world::world->getBall();
+
+    // if there is a robot to harass
+    if (targetRobotsToHarass[myIndex]) {
+        auto robotToHarass = targetRobotsToHarass[myIndex];
+        if (robotToHarass->hasBall()) {
+            return (ball->pos - thisRobot->pos).toAngle();
+        }
+        else {
+            return (robotToHarass->pos - thisRobot->pos).toAngle();
+        }
+    }
+    // else, aim towards the ball
+    return (ball->pos - thisRobot->pos).toAngle();
+}
+
+HarassRobotCoach::HarassTarget HarassRobotCoach::findRobotToHarass(const RobotPtr &thisRobot, int &myIndex,
+        bool goAfterBall) {
+    HarassTarget harassTarget;
```

Comment:
this function should be split into three smaller ones.

Reply:
extracted one function, but I think extracting more doesn't make it nicer